File: Backend/app/Domain/Bersama/Nilai/Kuantitas.php
<?php

declare(strict_types=1);

namespace App\Domain\Bersama\Nilai;

use Brick\Math\BigDecimal;
use Brick\Math\BigNumber;
use Brick\Math\RoundingMode;
use InvalidArgumentException;
use JsonSerializable;
use Stringable;

/**
 * Jumlah barang dengan presisi tetap 4 desimal (kg, meter, liter) (PRD §15.1, CLAUDE.md #7).
 */
final readonly class Kuantitas implements JsonSerializable, Stringable
{
    public const SKALA = 4;

    private function __construct(private BigDecimal $nilai) {}

    public static function Dari(BigNumber|int|string $nilai): self
    {
        $desimal = BigDecimal::of($nilai);

        if ($desimal->getScale() > self::SKALA && ! $desimal->toScale(self::SKALA, RoundingMode::Down)->isEqualTo($desimal)) {
            throw new InvalidArgumentException('Kuantitas maksimal '.self::SKALA." desimal: {$nilai}");
        }

        return new self($desimal->toScale(self::SKALA));
    }

    public static function Nol(): self
    {
        return new self(BigDecimal::zero()->toScale(self::SKALA));
    }

    public function Tambah(self $lain): self
    {
        return new self($this->nilai->plus($lain->nilai));
    }

    public function Kurangi(self $lain): self
    {
        return new self($this->nilai->minus($lain->nilai));
    }

    /** Konversi satuan, misal 2 dus × 12 = 24 pcs. */
    public function Kali(BigNumber|int|string $faktor, RoundingMode $mode = RoundingMode::HalfUp): self
    {
        return new self($this->nilai->multipliedBy($faktor)->toScale(self::SKALA, $mode));
    }

    public function Negasi(): self
    {
        return new self($this->nilai->negated());
    }

    public function Bandingkan(self $lain): int
    {
        return $this->nilai->compareTo($lain->nilai);
    }

    public function SamaDengan(self $lain): bool
    {
        return $this->nilai->isEqualTo($lain->nilai);
    }

    public function BernilaiNegatif(): bool
    {
        return $this->nilai->isNegative();
    }

    public function KeDesimal(): BigDecimal
    {
        return $this->nilai;
    }

    public function KeString(): string
    {
        return (string) $this->nilai;
    }

    public function jsonSerialize(): string
    {
        return $this->KeString();
    }

    public function __toString(): string
    {
        return $this->KeString();
    }
}

File: Paket/Inti/lib/Nilai/Kuantitas.dart
import 'package:decimal/decimal.dart';

import 'ModePembulatan.dart';

/// Jumlah barang/bahan dengan presisi tetap 4 desimal (PRD §8 F-05, §15.1, CLAUDE.md #7).
///
/// Padanan `App\Domain\Bersama\Nilai\Kuantitas` di Backend.
final class Kuantitas implements Comparable<Kuantitas> {
  const Kuantitas._(this._nilai);

  static const int skala = 4;

  final Decimal _nilai;

  /// Dari string desimal, misal `"2"` atau `"0.125"`.
  static Kuantitas Dari(String nilai) => DariDesimal(Decimal.parse(nilai));

  static Kuantitas DariBulat(int jumlah) => Kuantitas._(Decimal.fromInt(jumlah));

  static Kuantitas DariDesimal(Decimal nilai) {
    if (nilai.scale > skala) {
      throw ArgumentError.value(nilai.toString(), 'nilai', 'Kuantitas maksimal $skala desimal');
    }
    return Kuantitas._(nilai);
  }

  static Kuantitas Nol() => Kuantitas._(Decimal.zero);

  Kuantitas Tambah(Kuantitas lain) => Kuantitas._(_nilai + lain._nilai);

  Kuantitas Kurangi(Kuantitas lain) => Kuantitas._(_nilai - lain._nilai);

  /// Mengalikan (misal konversi satuan) lalu membulatkan ke 4 desimal dengan mode yang disebut eksplisit.
  Kuantitas Kali(Decimal faktor, {ModePembulatan mode = ModePembulatan.SetengahMenjauhiNol}) =>
      Kuantitas._(BulatkanKeSkala(_nilai * faktor, skala, mode));

  Kuantitas Negasi() => Kuantitas._(-_nilai);

  /// Hasil selalu -1, 0, atau 1 (sama dengan `BigDecimal::compareTo` di Backend).
  int Bandingkan(Kuantitas lain) => _nilai.compareTo(lain._nilai).sign;

  bool SamaDengan(Kuantitas lain) => _nilai == lain._nilai;

  bool BernilaiNol() => _nilai == Decimal.zero;

  bool BernilaiNegatif() => _nilai < Decimal.zero;

  Decimal KeDesimal() => _nilai;

  /// Format penyimpanan & API: string desimal, misal `"2.0000"`.
  String KeString() => _nilai.toStringAsFixed(skala);

  String toJson() => KeString();

  @override
  int compareTo(Kuantitas other) => Bandingkan(other);

  @override
  bool operator ==(Object other) => other is Kuantitas && SamaDengan(other);

  @override
  int get hashCode => _nilai.hashCode;

  @override
  String toString() => KeString();
}
